README.md
## abaculus
a small block of stone, tile, glass, or other material used in the construction of a mosaic

or,

a library for creating static maps from tiles based on center or corner lng,lat coordinates.
Uses node-mapnik to stitch tiles together.

[![Build Status](https://travis-ci.org/mapbox/abaculus.svg?branch=master)](https://travis-ci.org/mapbox/abaculus)


Looking to create high res images of maps? Abaculus was written for use in [Mapbox Studio](http://github.com/mapbox/mapbox-studio) and you can use Mapbox Studio to create and export high resolution images -- see [https://www.mapbox.com/guides/print/](https://www.mapbox.com/guides/print/) for more information. You can even use [that utility from the command line](https://github.com/mapbox/mapbox-studio/issues/1024#issuecomment-63535433).

### usage

Usage and example formatting below, or see the [tests](https://github.com/mapbox/abaculus/blob/master/test/test.js#L158-L204) for a more robust example with a `getTile` function.

#### input:
`scale`: integer between 1-4 and sets resolution (`scale: 1` is 72dpi, `scale: 4`, is 288dpi)

`zoom`: zoom level

`[w, s, e, n]`: the bounding box for the west (lat val), south (lng val), east (lat val), north (lng val) for the desired area

`x`: longitude coordinate

`y`: latitude coordinate

`width` and `height`: desired pixel bounds for a map with a center coordinate. Will be multiplied by scale to maintain resolution.

`format` (optional): `png` or `jpeg`, default is `png`.

`quality` (optional): when used with `jpeg` format, accepts 1-100 and defaults to 80. when used with `png` format, accepts 2-256 (# of colors to reduce the image to) and defaults to none.

`tileSize` (optional, defaults to `256`): Specifies input size of tiles used in `getTile` function.

`getTile`: a function that returns a tile buffer (png or otherwise) and headers given `z`, `x`, `y`, and a callback, such as from [tilelive-vector](https://github.com/mapbox/tilelive-vector/blob/master/index.js#L119-L218) or this [test function](https://github.com/mapbox/abaculus/blob/master/test/test.js#L184-L204).

`limit` (optional): max width or height of generated image in pixels. Default is `19008`.

```javascript
// Calculate image bounds from W,S,E,N bounding box.
var params = {
	zoom: {zoom},
	scale: {scale}
    bbox: [{w}, {s}, {e}, {n}],
    format: {format},
    quality: {quality},
	tileSize: {tileSize},
    getTile: function(z, x, y, callback){
		// do something
		return callback(null, buffer, headers);
	},
	limit: {limit}
};
```
or
```javascript
// Calculate image bounds from center lng,lat coordinates and
// pixel dimensions of final image (will be multipled by scale).
var params = {
	zoom: {zoom},
	scale: {scale}
    center: {
    	x: {x},
    	y: {y},
    	w: {width},
    	h: {height}
    },
    format: {format},
    quality: {quality},
	tileSize: {tileSize},
    getTile: function(z,x,y, callback){
		// do something
		return callback(null, buffer, headers);
	},
	limit: {limit}
};
```
#### usage:
``` javascript
abaculus(params, function(err, image, headers){
	if (err) return err;
	// do something with image
});
```

#### output:
an image of desired resolution for the selected area.


changelog.md
# CHANGELOG

# v3.1.1

Fixes a bug when calculating the center for sizes > 1

# v3.1.0

* Update mocha to 6.x
* Update mapnik to 4.2.1
* Update d3-queue to 3.0.7
* Add travis builds for node 6,8,10 latest lte releases


# v3.0.0

* Updates to mapnik 3.7.0
* Drops windows support

# v2.1.0

* Switches sphericalmercator to use mapbox repo
* Updates mapnik to 3.6.0

# v2.0.3

* Fixes a series of small bugs 

# v2.0.0

* Removes support for Node version `v0.12.x`
* Adds new optional `tileSize`. This is the native tileSize in pixels of the input tiles. If not provided, the API will default to `256`.


test/test.js
var assert = require('assert');
var printer = require('../');
var fs = require('fs');
var path = require('path');
var mapnik = require('mapnik');

// defaults
var zoom = 5,
    scale = 4,
    x = 4096,
    y = 4096,
    quality = 256,
    format = 'png',
    limit = 19008;


// fixtures
var tiles = fs.readdirSync(path.resolve(__dirname + '/fixtures/')).reduce(function(memo, basename) {
    var key = basename.split('.').slice(0, 4).join('.');
    memo[key] = fs.readFileSync(path.resolve(__dirname + '/fixtures/' + basename));
    return memo;
}, {});

describe('Get center from bbox', function() {
    it('should fail if (x1, y1) and (x2,y2) are equal', function() {
        var bbox = [0, 0, 0, 0];

        assert.throws( function() {
            printer.coordsFromBbox(zoom, scale, bbox, limit);
        }, /Incorrect coordinates/);
    });
    it('should fail if the image is too large', function() {
        var bbox = [-60, -60, 60, 60];

        assert.throws( function() {
            printer.coordsFromBbox(7, 2, bbox, limit);
        }, /Desired image is too large./);
    });
    it('should return the correct coordinates', function() {
        var bbox = [-60, -60, 60, 60];

        var center = printer.coordsFromBbox(zoom, scale, bbox, limit);
        assert.deepEqual(center.w, 10920);
        assert.deepEqual(center.h, 13736);
        assert.deepEqual(center.x, x);
        assert.deepEqual(center.y, y);
    });
});

describe('get coordinates from center', function() {
    it('should should fail if the image is too large', function() {
        var center = {
            x: 0,
            y: 0,
            w: 4752,
            h: 4752
        };
        assert.throws( function() {
            printer.coordsFromCenter(zoom, scale, center, limit);
        }, /Desired image is too large./);
    });
    it('should return correct origin coords', function() {
        var center = {
            x: 0,
            y: 20,
            w: 800,
            h: 800
        };
        center = printer.coordsFromCenter(zoom, scale, center, limit);
        assert.equal(center.x, x);
        assert.equal(center.y, 3631);
    });
    it('should return correct origin coords for negative y', function() {
        var zoom = 2,
            center = {
                x: 39,
                y: -14,
                w: 1000,
                h: 1000
            };
        center = printer.coordsFromCenter(zoom, scale, center, limit);
        assert.equal(center.x, 623);
        assert.equal(center.y, 552);
    });
});

describe('create list of tile coordinates', function() {
    it('should return a tiles object with correct coords', function() {
        var zoom = 5,
            scale = 4,
            width = 1824,
            height = 1832,
            center = { x: 4096, y: 4096, w: width, h: height };

        var expectedCoords = {
            tiles: [
                { z: zoom, x: 15, y: 15, px: -112, py: -108 },
                { z: zoom, x: 15, y: 16, px: -112, py: 916 },
                { z: zoom, x: 16, y: 15, px: 912, py: -108 },
                { z: zoom, x: 16, y: 16, px: 912, py: 916 }
            ],
            dimensions: { x: width, y: height },
            center: { row: 16, column: 16, zoom: zoom },
            scale: scale
        };
        var coords = printer.tileList(zoom, scale, center);
        assert.deepEqual(JSON.stringify(coords), JSON.stringify(expectedCoords));
    });

    it('should return a tiles object with correct coords when image exceeds y coords', function() {
        var zoom = 2,
            scale = 1,
            width = 1000,
            height = 1000,
            center = {x: 623, y: 552, w: width, h: height};

        var expectedCoords = {
            tiles: [
                { z: zoom, x: 0, y: 0, px: -123, py: -52 },
                { z: zoom, x: 0, y: 1, px: -123, py: 204 },
                { z: zoom, x: 0, y: 2, px: -123, py: 460 },
                { z: zoom, x: 0, y: 3, px: -123, py: 716 },
                { z: zoom, x: 1, y: 0, px:  133, py: -52 },
                { z: zoom, x: 1, y: 1, px:  133, py: 204 },
                { z: zoom, x: 1, y: 2, px:  133, py: 460 },
                { z: zoom, x: 1, y: 3, px:  133, py: 716 },
                { z: zoom, x: 2, y: 0, px:  389, py: -52 },
                { z: zoom, x: 2, y: 1, px:  389, py: 204 },
                { z: zoom, x: 2, y: 2, px:  389, py: 460 },
                { z: zoom, x: 2, y: 3, px:  389, py: 716 },
                { z: zoom, x: 3, y: 0, px:  645, py: -52 },
                { z: zoom, x: 3, y: 1, px:  645, py: 204 },
                { z: zoom, x: 3, y: 2, px:  645, py: 460 },
                { z: zoom, x: 3, y: 3, px:  645, py: 716 },
                { z: zoom, x: 0, y: 0, px:  901, py: -52 },
                { z: zoom, x: 0, y: 1, px:  901, py: 204 },
                { z: zoom, x: 0, y: 2, px:  901, py: 460 },
                { z: zoom, x: 0, y: 3, px:  901, py: 716 }
            ],
            dimensions: {x: width, y: height},
            center: {row: 2, column: 2, zoom: zoom},
            scale: scale
        };
        var coords = printer.tileList(zoom, scale, center);
        assert.deepEqual(JSON.stringify(coords), JSON.stringify(expectedCoords));
    });
    it('should avoid extranious images', function() {
        var zoom = 11,
            scale = 1,
            width = 256,
            height = 256,
            center = { w: width, h: height, x: 118912, y: 214912 };

        var expectedCoords = {
            tiles: [
                { z: 11, x: 464, y: 839, px: 0, py: 0},
                // {z: 11, x: 464, y: 840, px: 0, py: 256 },
                // { z: 11, x: 465, y: 839, px: 256, py: 0 },
                // { z: 11, x: 465, y: 840, px: 256, py: 256 }
            ],
            dimensions: {x: width, y: height},
            center: { row: 839, column: 464, zoom: 11 },
            scale: scale
        };
        var coords = printer.tileList(zoom, scale, center);
        assert.deepEqual(JSON.stringify(coords), JSON.stringify(expectedCoords));
    });
    it('should return a tiles object with correct coords when image is much bigger than world', function() {
        var zoom = 1,
            scale = 1,
            width = 2000,
            height = 2100,
            center = {x: 100, y: 100, w: width, h: height};

        var expectedCoords = {
            tiles: [
                {z: zoom, x: 0, y: 0, px: -124, py: 950},
                {z: zoom, x: 0, y: 1, px: -124, py: 1206},
                {z: zoom, x: 1, y: 0, px: 132, py: 950},
                {z: zoom, x: 1, y: 1, px: 132, py: 1206},
                {z: zoom, x: 0, y: 0, px: 388, py: 950},
                {z: zoom, x: 0, y: 1, px: 388, py: 1206},
                {z: zoom, x: 1, y: 0, px: 644, py: 950},
                {z: zoom, x: 1, y: 1, px: 644, py: 1206},
                {z: zoom, x: 0, y: 0, px: 900, py: 950},
                {z: zoom, x: 0, y: 1, px: 900, py: 1206},
                {z: zoom, x: 1, y: 0, px: 1156, py: 950},
                {z: zoom, x: 1, y: 1, px: 1156, py: 1206},
                {z: zoom, x: 0, y: 0, px: 1412, py: 950},
                {z: zoom, x: 0, y: 1, px: 1412, py: 1206},
                {z: zoom, x: 1, y: 0, px: 1668, py: 950},
                {z: zoom, x: 1, y: 1, px: 1668, py: 1206},
                {z: zoom, x: 0, y: 0, px: 1924, py: 950},
                {z: zoom, x: 0, y: 1, px: 1924, py: 1206}
            ],
            dimensions: {x: width, y: height},
            center: {row: 0, column: 0, zoom: zoom},
            scale: scale
        };
        var coords = printer.tileList(zoom, scale, center);
        assert.deepEqual(JSON.stringify(coords), JSON.stringify(expectedCoords));
    });
});

[256, 512, 1024].forEach(function(size) {
    describe('stitch tiles into single png', function() {
        var expectedCoords = {
            tiles: [
                { z: 1, x: 0, y: 0, px: 0, py: 0 },
                { z: 1, x: 0, y: 1, px: 0, py: size },
                { z: 1, x: 1, y: 0, px: size, py: 0 },
                { z: 1, x: 1, y: 1, px: size, py: size }
            ],
            dimensions: {
                x: size * 2,
                y: size * 2
            },
            center: { row: 1, column: 1, zoom: 1 },
            scale: 1,
            tileSize: size
        };

        it('should fail if no coordinates object', function(done) {
            printer.stitchTiles(null, format, quality, function() {}, function(err) {
                assert.equal(err.message, 'No coords object.');
                done();
            });
        });

        it('should return tiles and stitch them together', function(done) {
            var expectedImage = fs.readFileSync(path.resolve(__dirname + '/expected/expected.' + size + '.png'));

            printer.stitchTiles(expectedCoords, format, quality, getTileTest, function(err, image, header) {
                fs.writeFile(__dirname + '/outputs/expected.' + size + '.png', image, function(err){
                    checkImage(image, expectedImage);
                    done();
                });
            });
        });
    });

    describe('run entire function', function() {
        it('stitches images with a center coordinate', function(done) {
            var expectedImage = fs.readFileSync(path.resolve(__dirname + '/expected/center.' + size + '.png'));

            var params = {
                zoom: 1,
                scale: 1,
                center: {
                    x: 0,
                    y: 0,
                    w: 200,
                    h: 200
                },
                format: 'png',
                quality: 50,
                tileSize: size,
                getTile: getTileTest
            };

            printer(params, function(err, image) {
                assert.equal(err, null);

                fs.writeFile(__dirname + '/outputs/center.' + size + '.png', image, function(err){
                    assert.equal(err, null);
                    console.log('\tVisually check image at '+ __dirname + '/outputs/center.' + size + '.png');

                    // byte by byte check of image:
                    checkImage(image, expectedImage);
                    done();
                });
            });
        });

        it('stitches images with a wsen bbox', function(done) {
            var expectedImage = fs.readFileSync(path.resolve(__dirname + '/expected/bbox.' + size + '.png'));

            var params = {
                zoom: 1,
                scale: 1,
                bbox: [-140, -80, 140, 80],
                format: 'png',
                quality: 50,
                tileSize: size,
                getTile: getTileTest
            };

            printer(params, function(err, image, headers) {
                assert.equal(err, null);
                fs.writeFile(__dirname + '/outputs/bbox.' + size + '.png', image, function(err){
                    assert.equal(err, null);
                    console.log('\tVisually check image at '+ __dirname + '/outputs/bbox.'+ size +'.png');

                    // byte by byte check of image:
                    checkImage(image, expectedImage);
                    done();
                });
            });
        })
    });

    // This approximates a tilelive's getTile function
    // (https://github.com/mapbox/tilelive-vector/blob/master/index.js#L119-L218)
    // by loading a series of local png tiles
    // and returning the tile requested with the x, y, & z,
    // parameters along with the appropriate headers
    function getTileTest(z, x, y, callback) {
        var key = [z, x, y, size].join('.');

        // Headers.
        var headers = {
            'Last-Modified': new Date().toUTCString(),
            'ETag':'73f12a518adef759138c142865287a18',
            'Content-Type':'application/x-protobuf'
        };

        if (!tiles[key]) {
            return callback(new Error('Tile does not exist'));
        } else {
            return callback(null, tiles[key], headers);
        }
    }

    function checkImage(actual, expected) {
        actual = new mapnik.Image.fromBytes(actual);
        expected = new mapnik.Image.fromBytes(expected);
        var max_diff_pixels = 0;
        var compare_alpha = true;
        var threshold = 16;
        var diff_pixels = actual.compare(expected, {
            threshold: threshold,
            alpha: compare_alpha
        });
        if (diff_pixels > max_diff_pixels) {
            expected.save('test/outputs/center.fail.png');
        }
        assert.equal(max_diff_pixels, diff_pixels);
    }

});


index.js
var SphericalMercator = require('@mapbox/sphericalmercator');
var queue = require('d3-queue').queue;
var blend = require('mapnik').blend;
var crypto = require('crypto');

module.exports = abaculus;

function abaculus(arg, callback) {
    var z = arg.zoom || 0,
        s = arg.scale || 1,
        center = arg.center || null,
        bbox = arg.bbox || null,
        getTile = arg.getTile || null,
        format = arg.format || 'png',
        quality = arg.quality || null,
        limit = arg.limit || 19008,
        tileSize = arg.tileSize || 256;

    if (!getTile) return callback(new Error('Invalid function for getting tiles'));

    if (center) {
        // get center coordinates in px from lng,lat
        center = abaculus.coordsFromCenter(z, s, center, limit, tileSize);
    } else if (bbox) {
        // get center coordinates in px from [w,s,e,n] bbox
        center = abaculus.coordsFromBbox(z, s, bbox, limit, tileSize);
    } else {
        return callback(new Error('No coordinates provided.'));
    }
    // generate list of tile coordinates center
    var coords = abaculus.tileList(z, s, center, tileSize);

    // get tiles based on coordinate list and stitch them together
    abaculus.stitchTiles(coords, format, quality, getTile, callback);
}

abaculus.coordsFromBbox = function(z, s, bbox, limit, tileSize) {
    var sm = new SphericalMercator({ size: tileSize * s });
    var topRight = sm.px([bbox[2], bbox[3]], z),
        bottomLeft = sm.px([bbox[0], bbox[1]], z);
    var center = {};
    center.w = topRight[0] - bottomLeft[0];
    center.h = bottomLeft[1] - topRight[1];

    if (center.w <= 0 || center.h <= 0) throw new Error('Incorrect coordinates');

    var origin = [topRight[0] - center.w / 2, topRight[1] + center.h / 2];
    center.x = origin[0];
    center.y = origin[1];
    center.w = Math.round(center.w * s);
    center.h = Math.round(center.h * s);

    if (center.w >= limit || center.h >= limit) throw new Error('Desired image is too large.');
    return center;
};

abaculus.coordsFromCenter = function(z, s, center, limit, tileSize) {
    var sm = new SphericalMercator({ size: tileSize * s });
    var origin = sm.px([center.x, center.y], z);
    center.x = origin[0];
    center.y = origin[1];
    center.w = Math.round(center.w * s);
    center.h = Math.round(center.h * s);

    if (center.w >= limit || center.h >= limit) throw new Error('Desired image is too large.');
    return center;
};

// Generate the zxy and px/py offsets needed for each tile in a static image.
// x, y are center coordinates in pixels
abaculus.tileList = function(z, s, center, tileSize) {
    var x = center.x,
        y = center.y,
        w = center.w,
        h = center.h;
    var dimensions = {x: w - 1, y: h - 1};
    var size = tileSize || 256;
    var ts = Math.floor(size * s);

    var centerCoordinate = {
        column: x / size,
        row: y / size,
        zoom: z
    };

    function pointCoordinate(point) {
        var coord = {
            column: centerCoordinate.column,
            row: centerCoordinate.row,
            zoom: centerCoordinate.zoom,
        };
        coord.column += (point.x - w / 2) / ts;
        coord.row += (point.y - h / 2) / ts;
        return coord;
    }

    function coordinatePoint(coord) {
        // Return an x, y point on the map image for a given coordinate.
        if (coord.zoom != z) coord = coord.zoomTo(z);
        return {
            x: w / 2 + ts * (coord.column - centerCoordinate.column),
            y: h / 2 + ts * (coord.row - centerCoordinate.row)
        };
    }

    function floorObj(obj) {
        return {
            row: Math.floor(obj.row),
            column: Math.floor(obj.column),
            zoom: obj.zoom
        };
    }

    var maxTilesInRow = Math.pow(2, z);
    var tl = floorObj(pointCoordinate({x: 0, y:0}));
    var br = floorObj(pointCoordinate(dimensions));
    var coords = {};
    coords.tiles = [];

    for (var column = tl.column; column <= br.column; column++) {
        for (var row = tl.row; row <= br.row; row++) {
            var c = {
                column: column,
                row: row,
                zoom: z,
            };
            var p = coordinatePoint(c);

            // Wrap tiles with negative coordinates.
            c.column = c.column % maxTilesInRow;
            if (c.column < 0) c.column = maxTilesInRow + c.column;

            if (c.row < 0 || c.row >= maxTilesInRow) continue;
            coords.tiles.push({
                z: c.zoom,
                x: c.column,
                y: c.row,
                px: Math.round(p.x),
                py: Math.round(p.y)
            });
        }
    }
    coords.dimensions = { x: w, y: h };
    coords.center = floorObj(centerCoordinate);
    coords.scale = s;

    return coords;
};

abaculus.stitchTiles = function(coords, format, quality, getTile, callback) {
    if (!coords) return callback(new Error('No coords object.'));
    var tileQueue = queue(32);
    var w = coords.dimensions.x,
        h = coords.dimensions.y,
        s = coords.scale,
        tiles = coords.tiles;

    tiles.forEach(function(t) {
        tileQueue.defer(function(z, x, y, px, py, done) {
            var cb = function(err, buffer, headers) {
                if (err) return done(err);
                done(err, {
                    buffer: buffer,
                    headers: headers,
                    x: px,
                    y: py,
                    reencode: true
                })
            };
            cb.scale = s;
            cb.format = format;
            // getTile is a function that returns
            // a tile given z, x, y, & callback
            getTile(z, x, y, cb);
        }, t.z, t.x, t.y, t.px, t.py);
    });

    function tileQueueFinish(err, data) {
        if (err) return callback(err);
        if (!data) return callback(new Error('No tiles to stitch.'));
        var headers = [];
        data.forEach(function(d) {
            headers.push(d.headers);
        });

        blend(data, {
            format: format,
            quality: quality,
            width: w,
            height: h,
            reencode: true
        }, function(err, buffer) {
            if (err) return callback(err);
            callback(null, buffer, headerReduce(headers, format));
        });
    }

    tileQueue.awaitAll(tileQueueFinish);
};

// Calculate TTL from newest (max mtime) layer.
function headerReduce(headers, format) {
    var minmtime = new Date('Sun, 23 Feb 2014 18:00:00 UTC');
    var composed = {};

    composed['Cache-Control'] = 'max-age=3600';

    switch (format) {
    case 'vector.pbf':
        composed['Content-Type'] = 'application/x-protobuf';
        composed['Content-Encoding'] = 'deflate';
        break;
    case 'jpeg':
        composed['Content-Type'] = 'image/jpeg';
        break;
    case 'png':
        composed['Content-Type'] = 'image/png';
        break;
    }

    var times = headers.reduce(function(memo, h) {
        if (!h) return memo;
        for (var k in h) if (k.toLowerCase() === 'last-modified') {
            memo.push(new Date(h[k]));
            return memo;
        }
        return memo;
    }, []);
    if (!times.length) {
        times.push(new Date());
    } else {
        times.push(minmtime);
    }
    composed['Last-Modified'] = (new Date(Math.max.apply(Math, times))).toUTCString();

    var etag = headers.reduce(function(memo, h) {
        if (!h) return memo;
        for (var k in h) if (k.toLowerCase() === 'etag') {
            memo.push(h[k]);
            return memo;
        }
        return memo;
    }, []);
    if (!etag.length) {
        composed['ETag'] = '"' + crypto.createHash('md5').update(composed['Last-Modified']).digest('hex') + '"';
    } else {
        composed['ETag'] = etag.length === 1 ? etag[0] : '"' + crypto.createHash('md5').update(etag.join(',')).digest('hex') + '"';
    }

    return composed;
}
